src/driver-steelseries.c

```diff
@@ -234,20 +236,35 @@ steelseries_read_settings(struct ratbag_device *device)
 	if (ret < 0)
 		return ret;
 
-	if (device_version == 2) {
+	dpirange = ratbag_device_data_steelseries_get_dpi_range(device->data);
+
+	if (device_version == 2 || device_version == 4) {
 		active_resolution = buf[1] - 1;
 		ratbag_device_for_each_profile(device, profile) {
 			ratbag_profile_for_each_resolution(profile, resolution) {
 				resolution->is_active = resolution->index == active_resolution;
 
-				resolution->dpi_x = 100 * (1 +  buf[2 + resolution->index*2]);
+				if (device_version == 4) {
+					resolution->dpi_x = dpirange->step * get_unaligned_le_u16(&buf[3 + 2 * resolution->index]);
+				} else {
+					resolution->dpi_x = 100 * (1 +  buf[2 + resolution->index*2]);
+				}
 				resolution->dpi_y = resolution->dpi_x;
 			}
 
 			ratbag_profile_for_each_led(profile, led) {
-				led->color.red = buf[6 + led->index * 3];
-				led->color.green = buf[7 + led->index * 3];
-				led->color.blue = buf[8 + led->index * 3];
+				if (device_version == 4) {
+					led->color.red = buf[12 + led->index * 3];
+					led->color.green = buf[13 + led->index * 3];
+					led->color.blue = buf[14 + led->index * 3];
+				} else {
+					led->color.red = buf[6 + led->index * 3];
+					led->color.green = buf[7 + led->index * 3];
+					led->color.blue = buf[8 + led->index * 3];
+				}
+
+				if (!led->color.red && !led->color.green && !led->color.blue)
+					led->mode = RATBAG_LED_OFF;
```

Comment:
This bit seems unrelated. Please split it into a different commit if you can :sweat_smile: